src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,435 @@
+			 * function proceed
+			 */
+			zephyr_ll_task_done(sch, task);
+			k_sem_give(&pdata->sem);
+		} else if (state == SOF_TASK_STATE_COMPLETED) {
+			zephyr_ll_task_done(sch, task);
+		} else {
+			/*
+			 * task->state could've been changed to
+			 * SOF_TASK_STATE_CANCEL
+			 */
+			switch (task->state) {
+			case SOF_TASK_STATE_CANCEL:
+				zephyr_ll_task_done(sch, task);
+				break;
+			default:
+				/* reschedule */
+				list_item_del(&task->list);
+
+				zephyr_ll_task_insert_unlocked(sch, task);
+
+				task->start = sch->ll_domain->next_tick;
+			}
+		}
+
+		spin_unlock_irq(&sch->lock, flags);
+	}
+
+	notifier_event(sch, NOTIFIER_ID_LL_POST_RUN,
+		       NOTIFIER_TARGET_CORE_LOCAL, NULL, 0);
+}
+
+/* Called once for periodic tasks or multiple times for one-shot tasks */
+static int zephyr_ll_task_schedule(void *data, struct task *task, uint64_t start,
+				   uint64_t period)
+{
+	struct zephyr_ll *sch = data;
+	struct zephyr_ll_pdata *pdata;
+	struct task *task_iter;
+	struct list_item *list;
+	uint64_t delay = period ? period : start;
+	uint32_t flags;
+	int ret;
+
+	tr_info(&ll_tr, "task add %p %pU priority %d flags 0x%x", task, task->uid,
+		task->priority, task->flags);
+
+	irq_local_disable(flags);
+	if (!sch->n_tasks)
+		/*
+		 * If we're adding the first task to the scheduler, it is
+		 * possible that the scheduler has already run before on this
+		 * core and its thread has been aborted while holding the
+		 * spinlock, so we have to re-initialize it. Also note, that
+		 * the counter is decremented as the last operation in
+		 * zephyr_ll_task_done() before aborting the thread, so just
+		 * disabling local IRQs provides sufficient protection here.
+		 */
+		spinlock_init(&sch->lock);
+	irq_local_enable(flags);
+
+	spin_lock_irq(&sch->lock, flags);
+
+	/* check if task is already scheduled */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+
+		/* keep original start */
+		if (task_iter == task) {
+			spin_unlock_irq(&sch->lock, flags);
+			return 0;
+		}
+	}
+
+	pdata = task->priv_data;
+	pdata->period = period;
+
+	task->start = domain_time(sch) + delay;
+	if (sch->ll_domain->next_tick != UINT64_MAX &&
+	    sch->ll_domain->next_tick > task->start)
+		task->start = sch->ll_domain->next_tick;
+
+	zephyr_ll_task_insert_unlocked(sch, task);
+
+	sch->n_tasks++;
+
+	spin_unlock_irq(&sch->lock, flags);
+
+	ret = domain_register(sch->ll_domain, task, &zephyr_ll_run, sch);
+	if (ret < 0)
+		tr_err(&ll_tr, "zephyr_ll_task_schedule: cannot register domain %d",
+		       ret);
+
+	return 0;
+}
+
+/*
+ * This is synchronous - after this returns the object can be destroyed!
+ * Assertion: under Zephyr this is always called from a thread context!
+ */
+static int zephyr_ll_task_free(void *data, struct task *task)
+{
+	struct zephyr_ll *sch = data;
+	uint32_t flags;
+	struct zephyr_ll_pdata *pdata = task->priv_data;
+	bool must_wait;
+
+	if (k_is_in_isr())
+		k_panic();
```

Comment:
Lets not panic, just complain and return. Panic stops the trace.

Reply:
ok